### ISSUE TITLE
Append appears slow

### ISSUE THREAD

**Issue by cphillips83** [EventStore/EventStore#2227]
I actually wanted to continue the discussion on this issue here https://github.com/EventStore/EventStore/issues/1975 , while the findings were that the initial request was quite slow 1.2 seconds which should still be unacceptable, the resulting performance on subsequent appends is still unacceptable. 

I was testing the same thing as this other user and determined that, on avg, a single append to EventStore takes 8ms which is extremely slow in a append only design. I'm trying to determine where this is coming from and its a long shot from the proposed 10-20k the documents suggest you can achieve over TCP.

Are there benchmarks anywhere showing what raw through put should be considered in a single node setup with single thread vs multiple threads?

Is there a way to determine what could possibly be taking 8ms to append 50 bytes to a stream?

PC specs are fairly high end but running eventstore locally with no options passed.

**Comment by gregoryyoung** [comment]
The first call as I told you then was a JIT etc hit. This is not uncommon.

As for taking 8 ms. This is actually fairly ... normal. I can make it go much faster but that is including an fsync call to make sure the data is **actually on disk**. 

So in 8 ms you have 

the client starting the process
sending it to the node
the node receiving the message
parsing the message
determining the operation to occur
queuing the internal operation
doing the internal write operation
fsyncing the disk
responding internally
formatting messages to go back over TCP
sending data over TCP
receiving data from TCP
parsing the data
returning the information to the caller

The fsync alone is probably multiple ms but it depends on the disk.

**Comment by cphillips83** [comment]
I'm not going to lie, I had a hard time believing these numbers with the exception of fsync since I have no real experience with it. I did find quite a lot of data to suggest that its between 3-8ms on decent machines which if I had to guess is 99% of the entire Append time. All the other actions mentioned would be operating at microseconds on a loopback device and serializing/deserializing structured data is extremely fast. 

With all that being said, I wonder if it would be worth considering having a flag to control if it should wait for the commit or return that it acknowledged the commit. As long as things like projections and subscriptions do not fire until the commit is on the HDD, is there harm in allowing the client to continue on?

This would have to be a conscious decision, hence why a flag on Append to allow for immediate return. Using this model you would not want update your domain models in a third party DB and instead rely on subscriptions for eventual consistency exclusively.

I'm rambling at this point, but I suppose you could simulate that with your DDD model as well by using a thread pool to process the logic and immediately returning as if the command succeeded.

**Comment by gregoryyoung** [comment]
A one-way (best effort) append could be reasonable in some circumstances ...

I made a note about it.

**Comment by cphillips83** [comment]
Using Flush(true). Supposed it is calling ` Win32Native.FlushFileBuffers(this._handle);` which is the equivalent of fsync on windows?

>Another thing that makes measurements with it hard btw is that the entire queue is synced not just "your data". If there is a file copy etc other things might be getting synced.

Yea, I came across a cool article about FireFox using sqllite which ended up creating a fsync call per link click and basically coming to a complete halt when copying large files around.

**Comment by gregoryyoung** [comment]
You are calling FileStream.Flush(true) on windows?

**Comment by gregoryyoung** [comment]
btw the unbuffered io options might have some better (or possibly worse but more predictable!) and be worth trying ...

**Comment by cphillips83** [comment]
Yes, I'm writing data to the FS, could you elaborate unbuffered io options? Are you referring to a setting or argument to EventStore?

Btw, here is the test code if you care to see. My other machine is clearly doing some sort of trickery at the hardware level and I'm curious to see what drive/controller it has (can't tell until after work). If read about some controllers utilizing a "write on power loss" so that it can safely use dram cache and in an event of a power failure it will dump the remaining data to disk.

My work machine is showing 25-30ms which is painful to see....

```
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace fsync
{
    class Program
    {
        const int samples = 10;
        const int count = 10;

        static TimeSpan WriteToFile(string path, int dataSize)
        {

            if (File.Exists(path))
                File.Delete(path);
            var data = new byte[dataSize];
            for (var i = 0; i < dataSize; i++)
                data[i] = (byte)i;

            using (var fs = new FileStream(path, FileMode.Append, FileAccess.Write, FileShare.None, dataSize))
            {
                fs.SetLength(dataSize * count);
                var sw = Stopwatch.StartNew();
                for (var i = 0; i < count; i++)
                {
                    fs.Write(data, 0, dataSize);
                    fs.Flush(true);
                }
                sw.Stop();
                return sw.Elapsed;
            }
        }

        static void Main(string[] args)
        {
            for (var size = 512; size <= 2048; size += 512)
            {
                var smpls = new TimeSpan[samples];
                for (var i = 0; i < samples; i++)
                    smpls[i] = WriteToFile("P:\\test.txt", size);

                var avg = smpls.Average(x => x.TotalMilliseconds) / (float)count;
                Console.WriteLine($"Size: {size}, Avg: {avg}");
            }
        }
    }
}
```

**Comment by cphillips83** [comment]
> A one-way (best effort) append could be reasonable in some circumstances ...

I think it would be important to note that `best effort` should include that it will succeeded outside of hardware failures. For example, there should be internal memory cache of the object utilizing `ExpectedVersion` so that if 2 different `best effort` writes come in with different data that are both valid, the same rejection rules should apply on the memory cache version of the data so that if the server returns success, the write is guaranteed successful outside of hardware failures.
